File: f1sim/src/race/championship.ts
/**
 * Points across a season.
 *
 * A single race is a lap time; a championship is a reason to come back.
 * The table lives in localStorage under the same key the arcade game
 * used, so a season started there carries over rather than resetting
 * when the two halves became one.
 *
 * Scoring is deliberately dumb: it takes a finishing order and adds
 * points. It knows nothing about circuits, physics or sessions, which
 * is what lets it sit above both.
 */

/** The current F1 allocation, first to tenth. */
export const POINTS = [25, 18, 15, 12, 10, 8, 6, 4, 2, 1];

const KEY = 'f1go-season';

export type Table = Record<string, number>;

export const loadSeason = (): Table => {
  try {
    const raw = localStorage.getItem(KEY);
    return raw ? (JSON.parse(raw) as Table) : {};
  } catch {
    /* A corrupt or blocked store is not worth failing a race over. */
    return {};
  }
};

export const saveSeason = (table: Table): void => {
  try {
    localStorage.setItem(KEY, JSON.stringify(table));
  } catch {
    /* Private browsing, quota, or a user who has turned it off. The
       race still happened; it just will not be remembered. */
  }
};

export const resetSeason = (): void => {
  try {
    localStorage.removeItem(KEY);
  } catch {
    /* nothing to undo */
  }
};

/**
 * Score a finishing order into the season table.
 *
 * @returns the points the player took, for the results screen.
 */
export const score = (order: { name: string; player: boolean }[]): number => {
  const table = loadSeason();
  let earned = 0;
  order.forEach((row, i) => {
    const pts = POINTS[i] ?? 0;
    table[row.name] = (table[row.name] ?? 0) + pts;
    if (row.player) earned = pts;
  });
  saveSeason(table);
  return earned;
};

/** The table as a sorted list, for display. */
export const standings = (): { name: string; points: number }[] =>
  Object.entries(loadSeason())
    .map(([name, points]) => ({ name, points }))
    .sort((a, b) => b.points - a.points);

/* ------------------------------------------------------------------
   Best laps, kept per circuit. Also shares the arcade game's keys.
   ------------------------------------------------------------------ */
const bestKey = (circuitId: string): string => `f1go-best-${circuitId}`;

export const loadBest = (circuitId: string): number | null => {
  try {
    const v = Number.parseFloat(localStorage.getItem(bestKey(circuitId)) ?? '');
    return Number.isFinite(v) ? v : null;
  } catch {
    return null;
  }
};

export const saveBest = (circuitId: string, seconds: number): void => {
  try {
    localStorage.setItem(bestKey(circuitId), seconds.toFixed(3));
  } catch {
    /* see saveSeason */
  }
};

/** Records a lap and says whether it was a new best. */
export const recordLap = (circuitId: string, seconds: number): boolean => {
  const best = loadBest(circuitId);
  if (best !== null && seconds >= best) return false;
  saveBest(circuitId, seconds);
  return true;
};

File: unity/Packages/com.mumu.f1core/Runtime/Championship.cs
using System;
using System.Collections.Generic;

namespace MumuF1
{
    /// <summary>A driver and what they have scored.</summary>
    public struct Standing
    {
        public string Name;
        public int Points;
    }

    /// <summary>
    /// Somewhere to keep a season and a set of best laps.
    /// </summary>
    /// <remarks>
    /// An interface for the same reason <see cref="IGhostStore"/> is one: the
    /// reference reaches straight for <c>localStorage</c>, and this assembly
    /// has no host to reach for anything in.
    ///
    /// Every method may fail silently. A corrupt or blocked store is not worth
    /// failing a race over — the race still happened, it just will not be
    /// remembered.
    /// </remarks>
    public interface ISeasonStore
    {
        IDictionary<string, int> LoadSeason();
        void SaveSeason(IDictionary<string, int> table);
        void ResetSeason();

        double? LoadBest(string circuitId);
        void SaveBest(string circuitId, double seconds);
    }

    /// <summary>
    /// Points across a season, ported from
    /// <c>f1sim/src/race/championship.ts</c>.
    /// </summary>
    /// <remarks>
    /// A single race is a lap time; a championship is a reason to come back.
    ///
    /// Scoring is deliberately dumb: it takes a finishing order and adds
    /// points. It knows nothing about circuits, physics or sessions, which is
    /// what lets it sit above both.
    /// </remarks>
    public sealed class Championship
    {
        /// <summary>The current Formula 1 allocation, first to tenth.</summary>
        public static readonly int[] Points = { 25, 18, 15, 12, 10, 8, 6, 4, 2, 1 };

        private readonly ISeasonStore _store;

        public Championship(ISeasonStore store)
        {
            _store = store;
        }

        /// <summary>Score a finishing order into the season table.</summary>
        /// <returns>the points the player took, for the results screen.</returns>
        public int Score(IReadOnlyList<ClassificationRow> order)
        {
            IDictionary<string, int> table = _store.LoadSeason() ?? new Dictionary<string, int>();
            var earned = 0;

            for (var i = 0; i < order.Count; i++)
            {
                var points = i < Points.Length ? Points[i] : 0;

                table.TryGetValue(order[i].Name, out var had);
                table[order[i].Name] = had + points;

                if (order[i].Player) earned = points;
            }

            _store.SaveSeason(table);
            return earned;
        }

        /// <summary>The table as a sorted list, for display.</summary>
        public IReadOnlyList<Standing> Standings()
        {
            IDictionary<string, int> table = _store.LoadSeason() ?? new Dictionary<string, int>();

            var rows = new List<Standing>(table.Count);
            foreach (KeyValuePair<string, int> entry in table)
            {
                rows.Add(new Standing { Name = entry.Key, Points = entry.Value });
            }

            /* By points, then by name. The reference sorts on points alone and
               inherits its tie-break from the insertion order of a JavaScript
               object; a Dictionary has no order to inherit, so two drivers
               level on points would come out differently on every run. Name is
               arbitrary, and being arbitrary once is the point. */
            rows.Sort((a, b) =>
            {
                var byPoints = b.Points.CompareTo(a.Points);
                return byPoints != 0
                    ? byPoints
                    : string.Compare(a.Name, b.Name, StringComparison.Ordinal);
            });

            return rows;
        }

        /// <summary>Records a lap and says whether it was a new best.</summary>
        public bool RecordLap(string circuitId, double seconds)
        {
            var best = _store.LoadBest(circuitId);
            if (best != null && seconds >= best.Value) return false;

            _store.SaveBest(circuitId, seconds);
            return true;
        }
    }
}
